### PR TITLE
Fix missing duplicate diagnostic for enum declarations

### LINKED ISSUE #67258: CS7038: Failed to emit module 'name': Unable to determine specific cause of the failure.
**Version Used**: 

17.6 P1 using .NET 8.0 P1

**Steps to Reproduce**:

1. `ISourceGenerator` created a full C# source file with a partial class definition, containing a full `enum` definition
2. a standard C# source file on top also defined the same enum in the same partial class (by accident, unintentional)

**Diagnostic Id**:

`CS7038`

**Expected Behavior**:

`CS0102: The type '<class>' already contains a definition for '<enum>'`

**Actual Behavior**:

`CS7038: Failed to emit module '<name>': Unable to determine specific cause of the failure.`

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #67258

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 1]
```diff
@@ -847,18 +847,18 @@ internal override ManagedKind GetManagedKind(ref CompoundUseSiteInfo<AssemblySym
 
         internal bool IsUnsafe => HasFlag(DeclarationModifiers.Unsafe);
 
-        internal SyntaxTree AssociatedSyntaxTree => declaration.Declarations[0].Location.SourceTree;
+        private SyntaxTree? AssociatedSyntaxTree => IsFileLocal ? declaration.Declarations[0].Location.SourceTree : null;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:850]: >AssociatedSyntaxTree [](http://example.com/codeflow?start=28&length=20)

It looks like this API deserves a comment

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs [hunk 2]
```diff
@@ -448,6 +448,69 @@ void symbolValidator(ModuleSymbol symbol)
...
+            // 1.cs(2,13): error CS0101: The namespace '<global namespace>' already contains a definition for 'E'
+            // public enum E { }
+            Diagnostic(ErrorCode.ERR_DuplicateNameInNS, "E").WithArguments("E", "<global namespace>").WithLocation(2, 13),
+            // 1.cs(6,17): error CS0102: The type 'C' already contains a definition for 'E'
+            //     public enum E { }
+            Diagnostic(ErrorCode.ERR_DuplicateNameInClass, "E").WithArguments("C", "E").WithLocation(6, 17));
+    }
+
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/67258")]
+    public void NonFileLocalDelegate_Duplicate()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs:494]: >NonFileLocalDelegate_Duplicate [](http://example.com/codeflow?start=16&length=30)

Is it worth testing scenarios with file local types? Or do we have existing tests covering that?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs [hunk 3]
```diff
@@ -448,6 +448,69 @@ void symbolValidator(ModuleSymbol symbol)
...
+";
+        var comp = CreateCompilation(new[] { source, source });
+        comp.VerifyEmitDiagnostics(
+            // 1.cs(2,22): error CS0101: The namespace '<global namespace>' already contains a definition for 'D'
+            // public delegate void D();
+            Diagnostic(ErrorCode.ERR_DuplicateNameInNS, "D").WithArguments("D", "<global namespace>").WithLocation(2, 22),
+            // 1.cs(6,26): error CS0102: The type 'C' already contains a definition for 'D'
+            //     public delegate void D();
+            Diagnostic(ErrorCode.ERR_DuplicateNameInClass, "D").WithArguments("C", "D").WithLocation(6, 26));
+    }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs:512]: Please include the tests I documented in the issue: [https://github.com/dotnet/roslyn/issues/67258#issuecomment-1517084197](https://github.com/dotnet/roslyn/issues/67258#issuecomment-1517084197)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs:512]: I included the one marked TODO2_2, did you also want the scenario marked TODO2 included?

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs:494]: I think duplicate file class is well tested but duplicate file enum and delegate is under-tested. I'll add a few here.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs:622]: That's good. Thanks

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

**Comment by RikkiGibson** [comment]
@jcouv for a second review

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/FileModifierTests.cs:622]: Which test did that end up being? I can't find that scenario (just a partial containing an enum, without any file-local modifier, included twice in the compilation)

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 2)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Pull request merged**